Emit "source-read" events for files read via the ``include`` directive

halldorfannar opened the pull request:
Subject: Extend Sphinx to emit "source-read" events for RST files coming in via `include` directive.

### Feature or Bugfix
- Bugfix

### Purpose
- Requested PR by @AA-Turner based on discussion on issue #10678 

### Detail
Changes are well documented - please consult the code. Added a thorough unit test to expose the issue and fix it.

### Relates
- Closes #10678

halldorfannar committed:
Add test that exposes issue 10678

halldorfannar committed:
Subclass docutils RSTParser to fix issue

Our unit test is now passing.

halldorfannar committed:
Improve test

halldorfannar committed:
Improve test and fix issues uncovered.

Fixes #10678

halldorfannar committed:
Add CHANGES entry

halldorfannar committed:
Address feedback from GitHub CI

halldorfannar committed:
More clean-up to appease the style gods

Why doesn't the project apply these changes automatically on the main branch?

halldorfannar committed:
Trailing whitespaces removed

halldorfannar committed:
Fix unit tests

These tests needed to change with the introduction
of the derived class `sphinx.parsers.RSTStateMachine`.

halldorfannar committed:
Fix formatting in CHANGE entry

halldorfannar committed:
Improve comment

Noting that docutils 0.18 introduced the marker.

picnixz commented on sphinx/parsers.py:
```diff
@@ -20,6 +19,32 @@
+
+    def insert_input(self, include_lines, path):
+        # First we need to combine the lines back into text so we can send it with the
+        # source-read event. In newer releases of docutils there are two lines at the end,
```
Which release are you referring to? (we are checking docutils 0.18 - 0.20, so which of these differ here specifically?) For future maintenance, is there somewhere in docutils where we should be careful about?

halldorfannar replied:
The change to affix the marker in docutils was fortunately already there in version 0.18, as you can see [here](https://sourceforge.net/p/docutils/code/HEAD/tree/tags/docutils-0.18/docutils/parsers/rst/directives/misc.py#l196).  So there won't be a compatibility issue with the version range you listed.

picnixz replied:
Good then. If you can point which docutils version introduced this feature, maybe you could update the comment by saying "In docutils x.y+ [...]" (at least, future maintainers wouldn't ask themselves "which docutils version?"). If it's too much work, let's put something like "at least since 0.18".

halldorfannar replied:
Yes, no problem.  The change to affix the include marker was introduced in 0.18.  I will update the comment with that information.

picnixz commented on sphinx/parsers.py:
```diff
@@ -20,6 +19,32 @@
+        # source-read event. In newer releases of docutils there are two lines at the end,
+        # that act as markers. We must preserve them and leave them out of the source-read
+        # event:
+        text = "\n".join(include_lines[:-2])
```
Is there a way to avoid having this join back ? like, does the docutils class expose the included lines cached as strings ?

halldorfannar replied:
I wish there was, but as you can see [here](https://sourceforge.net/p/docutils/code/HEAD/tree/tags/docutils-0.18/docutils/parsers/rst/directives/misc.py#l54) the text gets included, where options can dictate that only a chunk of the text is used.  See `start-line`/`end-line` and `start-after`/`end-before`.  The `rawtext` is unfortunately never assigned to a member variable so there isn't a way for us to access it. In order to make things easier in the future, then perhaps docutils maintainers could be persuaded to break their current implementation of the `run` method into two methods, something like:

* `extract_text`
* `parse_text`

And in docutils, the `Include.run` method would just call both of these in succession (passing results from the former to the latter).  In sphinx, we would reimplement `Include.run`  to call `extract_text`, then send the "source-read" event if the included file is RST, and pass the results along to `parse_text`.  I would argue that the `run` method in docutils is already quite large so breaking it up logically like suggested above is a good thing.

picnixz replied:
Yes, that would be a good idea. I actually find quite frustrating that docutils (or Sphinx actually) have some hardcoded implementations that are hard to monkey-patch without either rewriting the whole flow (which then depends on upstream changes). It'd be better if these implementations had (protected) methods that we could subclass when needed without having to bother about the rest of the implementation.

halldorfannar replied:
Agreed. I do think docutils would benefit from some refactoring to make customization easier and more robust.

picnixz commented on tests/test_directive_other.py:
```diff
@@ -148,3 +148,22 @@ def test_toctree_twice(app):
+
+    def source_read_handler(app, doc, source):
+        sources_reported[doc] = source[0]
+    app.connect("source-read", source_read_handler)
```
Btw, is there any issue if you have other `.. include::` directives inside an included document? You could also add a test to reflect the original issue, namely transforming the included content.

halldorfannar replied:
The test that I added does reflect the original issue: no "source-read" event generated for included RST files.  

I can absolutely add another test to transform the content, and include within an include file.

halldorfannar replied:
I have now added the test and you were indeed right that includes of includes were problematic with the sub-classing approach.  I have reverted to the patching solution which I used in the original extension I made for Sphinx.

halldorfannar committed:
Merge branch 'issue-10678' of https://github.com/halldorfannar/sphinx into issue-10678

halldorfannar committed:
Use :dudir: to refer to include directive

picnixz commented on CHANGES:
```diff
@@ -20,7 +20,7 @@ Features added
 ----------
 
-* #10678: Emit "source-read" events for RST files read in via `include` directive. 
+* #10678: Emit "source-read" events for RST files read in via ``include`` directive. 
```
You could use <code>:dudir:\`include\`</code>.

halldorfannar replied:
Thanks for the tip! Done.

halldorfannar committed:
Revert "Fix unit tests"

This reverts commit 4c727ceb7673afe74864736b30da44d4f655bdac.

halldorfannar committed:
Change approach

The subclassing will not work because docutils doesn't use Sphinx's rst
parser for processing the included
rst text.  This is a shame, and
strictly speaking an issue in Sphinx/docutils.
So we go back to monkey patching.
It works, the change is now better contained inside Sphinx Include directive.
Added a unit test to verify that
includes of includes are correctly
handled.

halldorfannar committed:
Fix test failure on Windows

halldorfannar committed:
Add type: ignore to work around known issue

halldorfannar committed:
Address flake8 issue

These are not part of my change but mysteriously now
cause failure in CI.

halldorfannar committed:
Merge branch '7.1.x' into issue-10678

halldorfannar committed:
Adjust CHANGES

created Unreleased section and
moved my change notes to it.

halldorfannar committed:
Add space to silence flake8

halldorfannar committed:
Merge branch '7.1.x' into issue-10678

AA-Turner committed:
Merge branch 'master' into issue-10678

# Conflicts:
#	CHANGES

picnixz commented on CHANGES:
```diff
@@ -36,6 +36,9 @@ Features added
   Patch by Rouslan Korneychuk.
 * 10938: doctest: Add :confval:`doctest_show_successes` option.
   Patch by Trey Hunner.
+* #10678: Emit "source-read" events for files read via
```
Let's order it by # (so I assume that it should come before 10938). At least, it will save some work for @AA-Turner when doing the next release

AA-Turner replied:
I tend to order it chronologically, by which change was introduced first -- but I'm inconsistent, and I do sort by other methods sometimes.

A

picnixz commented on sphinx/directives/other.py:
```diff
@@ -357,6 +359,30 @@ class Include(BaseInclude, SphinxDirective):
+        # when parsing RST text that comes in via Include directive.
+        def insert_input(include_lines, path):
+            # First we need to combine the lines back into text so we can send it with the
+            # source-read event. In docutils 0.18 and later, there are two lines at the end,
```
```suggestion
            # source-read event. In docutils 0.18 and later, there are two lines at the end
```

picnixz reviewed (approved):
Just wondering but do you think it's worth to actually have this behaviour disabled by default? because if there are a lot of included files that are actually not processed by the `source-read` event, then the overhead might not be negligible for very large projects (I don't know if large projects use `include` directives a lot or not). Maybe it's good to avoid slowing down builds (you essentially perform "join/splits" twice which probably increases your build time by 10-20% if you have a lot of files). If my question is actually a non-question, LGTM.

AA-Turner committed:
Merge branch 'master' into issue-10678

halldorfannar committed:
Update sphinx/directives/other.py

Co-authored-by: picnixz <10796600+picnixz@users.noreply.github.com>

AA-Turner committed:
Merge branch 'master' into issue-10678

# Conflicts:
#	CHANGES

AA-Turner committed:
Formatting

AA-Turner committed:
Merge branch 'master' into issue-10678

# Conflicts:
#	sphinx/directives/other.py

AA-Turner committed:
Only patch if there are source-read listeners
